### PR TITLE
feat(graphQL): add model check for chat, completion, embedding

### PR DESCRIPTION
**Pull request by zwpaper**
- backend for https://github.com/TabbyML/tabby/issues/3239
- Both local model and HTTP model tested

CleanShot 2024-10-28 at 15 40 22@2x

CleanShot 2024-10-28 at 15 41 18@2x

CleanShot 2024-10-28 at 15 41 47@2x

notes:
1. the `CompletionService` in tabby-inference is wrapped by `CodeGeneration`, and `CodeGeneration` is then wrapped by `CompletionService` struct in tabby service, we should use the CompltionService in tabby-inference, but it's a bit tricky according to the current implementation, so I returned the `CodeGeneration` from tabby-inference, and use it as the GraphQL context.
2. when failed to connect to the model, GraphQL request will fail, so I just return the latency as the success indication, it's `milliseconds` i32.
4. following the GraphQL convention, the enum is in SCREAMING_SNAKE_CASE by default

### CODE REVIEW

File: ee/tabby-schema/src/schema/mod.rs [hunk 1]
```diff
@@ -658,6 +668,25 @@ impl Query {
 
         Ok(SourceIdAccessPolicy { source_id, read })
     }
+
+    async fn test_model_connection(ctx: &Context) -> Result<ModelHealth> {
```
Comment (wsxiaoys) [inline ee/tabby-schema/src/schema/mod.rs:672]: adding argument (can be an enum) of `chat`, `completion`, `embedding`.

**Comment by zwpaper** [comment]
cc @liangfung

File: ee/tabby-schema/src/schema/mod.rs [hunk 2]
```diff
@@ -658,6 +673,50 @@ impl Query {
 
         Ok(SourceIdAccessPolicy { source_id, read })
     }
+
+    async fn test_model_connection(ctx: &Context, backend: ModelHealthBackend) -> Result<i32> {
```
Comment (wsxiaoys) [inline ee/tabby-schema/src/schema/mod.rs:677]: would it be better to create a new error type for Result (Result<i32, TestModelConnectionError>)? e.g we might wanna report different error type to user in future (e.g credential error, timeout, connection failed, ...)

**Review by wsxiaoys — COMMENTED** [review]
Otherwise LGTM

File: ee/tabby-schema/src/schema/mod.rs (hunk 2 above)
Comment (zwpaper) [reply ee/tabby-schema/src/schema/mod.rs]: All the GraphQL APIs return the CoreError, so I added a field `ModelConnectionFailed(String)` in CoreError

- return NotFound, when the model backend is not running
- return ModelConnectionFailed with error message, when failed to connect model

File: ee/tabby-webserver/src/service/mod.rs [hunk 3]
```diff
@@ -325,6 +341,8 @@ impl ServiceLocator for ArcServerContext {
 
 pub async fn create_service_locator(
     logger: Arc<dyn EventLogger>,
+    chat: Option<Arc<dyn ChatCompletionStream>>,
+    completion: Option<Arc<CodeGeneration>>,
```
Comment (wsxiaoys) [inline ee/tabby-webserver/src/service/mod.rs:345]:
```suggestion
    completion: Option<Arc<CompletionStream>>,
```

`CodeGeneration` is a wrapper interface - use CompletionStream directly.

File: ee/tabby-schema/src/schema/mod.rs [hunk 4]
```diff
@@ -188,6 +220,18 @@ async fn check_license(ctx: &Context, license_type: &[LicenseType]) -> Result<()
...
 
+#[derive(GraphQLEnum)]
+enum ModelHealthBackend {
+    Chat,
+    Completion,
+    Embedding,
+}
+
+#[derive(GraphQLObject, Debug, Clone)]
+struct ModelHealthResponse {
```
Comment (wsxiaoys) [inline ee/tabby-schema/src/schema/mod.rs:231]:
```suggestion
struct ModelBackendHealthInfo {
```

File: ee/tabby-schema/src/schema/mod.rs [hunk 5]
```diff
@@ -139,6 +153,24 @@ impl<S: ScalarValue> IntoFieldError<S> for CoreError {
     }
 }
 
+#[derive(thiserror::Error, Debug)]
+pub enum TestModelConnectionError {
+    #[error("{0}")]
+    CoreError(#[from] CoreError),
+
```
Comment (wsxiaoys) [inline ee/tabby-schema/src/schema/mod.rs:160]:
```suggestion
```

Consider adding an `Other(String)` error type instead.

File: ee/tabby-schema/src/schema/mod.rs [hunk 6]
```diff
@@ -139,6 +153,24 @@ impl<S: ScalarValue> IntoFieldError<S> for CoreError {
...
+    #[error("{0}")]
+    ModelConnectionFailed(String),
```
Comment (wsxiaoys) [inline ee/tabby-schema/src/schema/mod.rs:162]:
```suggestion
    FailedToConnect(String),
```

**Review by wsxiaoys — CHANGES_REQUESTED** [review]

**Pull request merged**